pkg/wekafs/dirvolume.go

```diff
@@ -0,0 +1,388 @@
+}
+
+//getInodeId used for obtaining the mount Path inode ID (to set quota on it later)
+func (v DirVolume) getInodeId(mountPath string) (uint64, error) {
+	glog.V(5).Infoln("Getting inode ID of volume", v.GetId(), "fullpath: ", v.getFullPath(mountPath))
+	fileInfo, err := os.Stat(v.getFullPath(mountPath))
+	if err != nil {
+		glog.Error(err)
+		return 0, err
+	}
+	stat, ok := fileInfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, errors.New(fmt.Sprintf("failed to obtain inodeId from %s", mountPath))
+	}
+	glog.V(5).Infoln("Inode ID of the volume", v.GetId(), "is", stat.Ino)
+	return stat.Ino, nil
+}
+
+func (v DirVolume) GetId() string {
+	return v.id
+}
+
+func (v DirVolume) CreateQuotaFromVolumeName(mountPath string, enforceCapacity *bool, capacityLimit uint64) (*apiclient.Quota, error) {
+	var quotaType apiclient.QuotaType
+	if enforceCapacity != nil {
+		if !*enforceCapacity {
+			quotaType = apiclient.QuotaTypeSoft
+		} else {
+			quotaType = apiclient.QuotaTypeHard
+		}
+	} else {
+		quotaType = apiclient.QuotaTypeDefault
+	}
+	glog.V(4).Infoln("Creating a quota for volume", v.GetId(), "capacity limit:", capacityLimit, "quota type:", quotaType)
+
+	fs, err := v.getFilesystemObj()
+	if err != nil {
+		return nil, err
+	}
+	inodeId, err := v.getInodeId(mountPath)
+	if err != nil {
+		return nil, errors.New("cannot set quota, could not find inode ID of the volume")
+	}
+	qr := apiclient.NewQuotaCreateRequest(*fs, inodeId, quotaType, capacityLimit)
+	q := &apiclient.Quota{}
+	if err := v.apiClient.CreateQuota(qr, q, true); err != nil {
+		return nil, err
+	}
+	glog.V(4).Infoln("Quota successfully set for volume", v.GetId())
+	return q, nil
+}
+
+func (v DirVolume) getQuota(mountPath string) (*apiclient.Quota, error) {
+	glog.V(4).Infoln("Getting existing quota for volume", v.GetId())
+	fs, err := v.getFilesystemObj()
+	if err != nil {
+		return nil, err
+	}
+	inodeId, err := v.getInodeId(mountPath)
+	if err != nil {
+		return nil, err
+	}
+	ret, err := v.apiClient.GetQuotaByFileSystemAndInode(fs, inodeId)
+	if ret != nil {
+		glog.V(4).Infoln("Successfully acquired existing quota for volume", v.GetId(), ret.GetQuotaType(), ret.GetCapacityLimit())
+	}
+	return ret, err
+}
+
+func (v DirVolume) getSizeFromQuota(mountPath string) (uint64, error) {
+	q, err := v.getQuota(mountPath)
+	if err != nil {
+		return 0, err
+	}
+	if q != nil {
+		return q.GetCapacityLimit(), nil
+	}
+	return 0, errors.New("could not fetch quota from API")
+}
+
+func (v DirVolume) getSizeFromXattr(mountPath string) (uint64, error) {
+	if capacityString, err := xattr.Get(v.getFullPath(mountPath), xattrCapacity); err == nil {
+		if capacity, err := strconv.ParseInt(string(capacityString), 10, 64); err == nil {
+			return uint64(capacity), nil
+		}
+		return 0, ErrBadXattrOnVolume
+	}
+	return 0, ErrNoXattrOnVolume
+}
+
+//getSizeFromDf used for GetCapacity, since we don't have secrets to check API
+func (v DirVolume) getSizeFromDf(mountPath string) (uint64, error) {
```

Comment:
looks like not needed

Reply:
Did this to support volume.GetCapacity() for ControllerServer.GetCapacity() but it can't be done anyway.